src/Lean/Meta/Tactic/Simp/Main.lean

```diff
@@ -235,9 +235,15 @@ inductive SimpLetCase where
   | nondep -- `let x := v; b` is equivalent to `(fun x => b) v`, and result type does not depend on `x`
 deriving Repr
 
-def getSimpLetCase (n : Name) (t : Expr) (b : Expr) : MetaM SimpLetCase := do
+def getSimpLetCase (n : Name) (t : Expr) (b : Expr) (nondep : Bool) : MetaM SimpLetCase := do
   withLocalDeclD n t fun x => do
     let bx := b.instantiate1 x
+    if nondep then
+      let bxType ← whnf (← inferType bx)
+      if (← dependsOn bxType x.fvarId!) then
```

Comment:
@kmill Could you please add a comment here explaining this part? We will forget.

Reply:
Added an explanation.

By the way, I have a followup PR that intends to replace this code. Brief preview: if `simp` comes across a `nondep := false` let, then it can run letToHave on it (which is slightly cheaper than `isTypeCorrect`, and it addresses the "TODO" comment in this function, by handling all lets simultaneously), and otherwise we can do the nondepDepVar/nondep determination on an entire telescope at once, using code similar to what you wrote for letFun.